browser/psst/psst_tab_web_contents_observer_browsertest.cc

```diff
@@ -0,0 +1,240 @@
+                        base::Base64UrlEncodePolicy::OMIT_PADDING,
+                        &encoded_destination);
+  const std::string query =
+      base::JoinString({"url", encoded_destination.c_str()}, "=");
+  GURL::Replacements replacement;
+  replacement.SetQueryStr(query);
+  return navigation_url.ReplaceComponents(replacement);
+}
+
+}  // namespace
+
+class MockComponentUpdateService
+    : public component_updater::MockComponentUpdateService {
+ public:
+  MockComponentUpdateService() = default;
+  ~MockComponentUpdateService() override = default;
+};
+class PsstTabWebContentsObserverBrowserTest : public PlatformBrowserTest {
+ public:
+  using LoadRulesTestCallback = base::MockCallback<base::OnceCallback<
+      void(const std::string& data, const std::vector<PsstRule>& rules)>>;
+
+  PsstTabWebContentsObserverBrowserTest()
+      : https_server_(net::EmbeddedTestServer::TYPE_HTTPS) {
+    feature_list_.InitAndEnableFeature(psst::features::kEnablePsst);
+  }
+
+  void SetUpOnMainThread() override {
+    PlatformBrowserTest::SetUpOnMainThread();
+    base::FilePath test_data_dir =
+        base::PathService::CheckedGet(base::DIR_SRC_TEST_DATA_ROOT);
+
+    base::PathService::Override(
+        component_updater::DIR_COMPONENT_PREINSTALLED,
+        test_data_dir.AppendASCII("brave/components/test/data/psst"));
+
+    https_server_.ServeFilesFromDirectory(
+        test_data_dir.AppendASCII("brave/test/data"));
+    https_server_.AddDefaultHandlers(GetChromeTestDataDir());
+    https_server_.SetSSLConfig(net::EmbeddedTestServer::CERT_TEST_NAMES);
+
+    host_resolver()->AddRule("*", "127.0.0.1");
+    ASSERT_TRUE(https_server_.Start());
+
+    SetUpPsstComponent();
+  }
+
+  virtual void SetUpPsstComponent() {
+    EXPECT_CALL(component_update_service_, RegisterComponent(testing::_))
+        .Times(1)
+        .WillOnce(testing::Return(true));
+    LoadRulesTestCallback mock_callback;
+    base::RunLoop run_loop;
+    PsstRuleRegistry::GetInstance()->SetOnLoadCallbackForTesting(
+        mock_callback.Get());
+    EXPECT_CALL(mock_callback, Run)
+        .Times(1)
+        .WillOnce(
+            [&](const std::string& data, const std::vector<PsstRule>& rules) {
+              EXPECT_FALSE(rules.empty());
+              EXPECT_FALSE(data.empty());
+              run_loop.Quit();
+            });
+
+    RegisterPsstComponent(&component_update_service_);
+    run_loop.Run();
+  }
+
+  PrefService* GetPrefs() { return browser()->profile()->GetPrefs(); }
+
+  net::EmbeddedTestServer& GetEmbeddedTestServer() { return https_server_; }
+
+  content::WebContents* web_contents() {
+    return chrome_test_utils::GetActiveWebContents(this);
+  }
+
+ protected:
+  net::EmbeddedTestServer https_server_;
+  base::test::ScopedFeatureList feature_list_;
+  MockComponentUpdateService component_update_service_;
+};
+
+// TESTS
+IN_PROC_BROWSER_TEST_F(PsstTabWebContentsObserverBrowserTest,
+                       DontStartScriptHandlerForSameDocument) {
```

Comment:
this is already covered in the unit tests. We don't want to duplicate tests here, we only need to test that everything works together as expected for typical cases